### PR TITLE
[Lazy init] Force fall back to slow init for composite models

### LINKED ISSUE #11704: [RAG] official facebook example code for RAG is not working anymore.
## Environment info
- `transformers` version: 4.6.0.dev0
- Platform: Linux-3.10.0-1127.10.1.el7.x86_64-x86_64-with-debian-buster-sid
- Python version: 3.6.13
- PyTorch version (GPU?): 1.8.1+cu102 (True)
- Tensorflow version (GPU?): 2.4.1 (False)
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help
rag: @patrickvonplaten, @lhoestq

Models:
RAG model

## Information

Model I am using RAG

The problem arises when using the official example scripts from https://huggingface.co/facebook/rag-sequence-nq, which I copied here:

```
from transformers import RagTokenizer, RagRetriever, RagSequenceForGeneration 

tokenizer = RagTokenizer.from_pretrained("facebook/rag-sequence-nq") 
retriever = RagRetriever.from_pretrained("facebook/rag-sequence-nq", index_name="exact", use_dummy_dataset=True) 
model = RagSequenceForGeneration.from_pretrained("facebook/rag-sequence-nq", retriever=retriever) 

input_dict = tokenizer.prepare_seq2seq_batch("how many countries are in europe", return_tensors="pt") 

generated = model.generate(input_ids=input_dict["input_ids"]) 
print(tokenizer.batch_decode(generated, skip_special_tokens=True)[0]) 

```

The tasks I am working on is:
I am trying to ran the sample code from above. Note that this same error occurs also when finetuning rag with the official code in transformers/examples/research_projects/rag/finetune_rag.sh

## To reproduce

1. run the sample code above 

The error is:
```
Traceback (most recent call last):
  File "prova_rag.py", line 5, in <module>
    model = RagSequenceForGeneration.from_pretrained("facebook/rag-sequence-nq", retriever=retriever, _fast_init=False)
  File "/nlu/users/giovanni_bonetta/transformers/src/transformers/modeling_utils.py", line 1208, in from_pretrained
    model, state_dict, pretrained_model_name_or_path
  File "/nlu/users/giovanni_bonetta/transformers/src/transformers/modeling_utils.py", line 1278, in _load_state_dict_into_model
    model._init_weights(module)
  File "/nlu/users/giovanni_bonetta/miniconda2/envs/hf_venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 948, in __getattr__
    type(self).__name__, name))
AttributeError: 'RagSequenceForGeneration' object has no attribute '_init_weights'
```

Looking at last commits i suppose that the error was introduced in "Pytorch - Lazy initialization of models #11471" a couple of weeks ago, where the line `model._init_weights(module)` was introduced.
## Expected behavior

It should initialize the model without errors.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Thanks to the great issue #11704 it was discovered that fast initialization currently breaks for all models whose `XXXPreTrainedModel` does not implement a `_init_weights` function and for which parts of the weights are missing when using `.from_pretrained(...)`. This includes essentially all composite models, being `Rag` and `EncoderDecoder`.

This PR does the vanilla fix of forcing those models to fall back on `_slow_init` since a better fix requires a careful re-design which is left for a future PR.

## Future PR
- [ ] Remove hacky `from_pretrained(...)` methods in RAG and EncoderDecoder
- [ ] Refactor the way "fast_init" calls `model._init_weights` for composite models. For Composite models, each part has to be called directly => 

```python
model.encoder._init_weigths(all_missing_keys_of_encoder)
model.decoder._init_weigths(all_missing_keys_of_decoder)
```

- [ ] Add more tests for RAG & EncoderDecoder

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py [hunk 1]
```diff
@@ -221,6 +221,13 @@ def get_output_embeddings(self):
     def set_output_embeddings(self, new_embeddings):
         return self.decoder.set_output_embeddings(new_embeddings)
 
+    @classmethod
+    def from_pretrained(cls, *args, **kwargs):
+        # At the moment fast initialization is not supported
+        # for composite models
+        kwargs["_fast_init"] = False
+        return super(EncoderDecoderModel, cls).from_pretrained(*args, **kwargs)
```
Comment (LysandreJik) [inline src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:229]:
```suggestion
        return super().from_pretrained(*args, **kwargs)
```

File: src/transformers/models/rag/modeling_rag.py [hunk 2]
```diff
@@ -232,6 +232,13 @@ class RagPreTrainedModel(PreTrainedModel):
     base_model_prefix = "rag"
     _keys_to_ignore_on_load_missing = [r"position_ids"]
 
+    @classmethod
+    def from_pretrained(cls, *args, **kwargs):
+        # At the moment fast initialization is not supported
+        # for composite models
+        kwargs["_fast_init"] = False
+        return super(RagPreTrainedModel, cls).from_pretrained(*args, **kwargs)
```
Comment (LysandreJik) [inline src/transformers/models/rag/modeling_rag.py:240]:
```suggestion
        return super().from_pretrained(*args, **kwargs)
```

**Review by LysandreJik — APPROVED** [review]
Eventually would be nice to add a test for this, as this is an issue that should have been caught by the tests!

Thanks for the hotfix.

**Review by sgugger — APPROVED** [review]
LGTM for a hotfix.

**Pull request merged**